### PR TITLE
Audio: Use generic saturation logic for improved efficiency

### PR DESCRIPTION
**Pull request by ShriramShastry**
Introduce a arithmetic bitwise saturation function operations across different integer sizes (8, 16, 24, 32 bits) in the 
processing block. 

- Replaced `if-else` checks with bitwise masks to handle overflow and
  underflow for int32, int24, int16, and int8 saturation functions.

- For each bit width, created masks by shifting the results of comparisons
  (x - `MIN_VAL`) and (`MAX_VAL` - x) to the respective overflow/underflow
  detection bits.

- Mask >> `31` or >> `63` results in either `0` (no overflow/underflow) or `-1`
  (overflow/underflow occurred).

- Applied bitwise operations to conditionally replace original values
  with boundary values (`MIN` or `MAX`) based on the computed masks.

- This approach avoids branching, improves efficiency, and ensures
  accurate saturation within specified bit-width limits.
  
**_HiFi3 Inline Summary_**
sat_int32: Decreased from 17 to 6-9 cycles; Bit true: Mixed; Gain: ~47.06% to 64.71%
sat_int24: Decreased from 7 to 5 cycles; Bit true: Yes; Gain: ~28.57%
sat_int16: Decreased from 13 to 7 cycles; Bit true: Yes; Gain: ~46.15%
sat_int8: Decreased from 3 to 1 cycle; Bit true: Yes; Gain: ~66.67%
 
**_HiFi4 Inline Summary_**
sat_int32: Decreased from 11 to 6-8 cycles; Bit true: Mixed; Gain: ~27.27% to 45.45%
sat_int24: No change (5 cycles); Bit true: Yes
sat_int16: Decreased from 9 to 7 cycles; Bit true: Yes; Gain: ~22.22%
sat_int8: Decreased from 3 to 1 cycle; Bit true: Yes; Gain: ~66.67%
 
_**HiFi5 Inline Summary**_
sat_int32: Decreased from 11 to 6-8 cycles; Bit true: Mixed; Gain: ~27.27% to 45.45%
sat_int24: No change (5 cycles); Bit true: Yes
sat_int16: Decreased from 9 to 7 cycles; Bit true: Yes; Gain: ~22.22%
sat_int8: Decreased from 3 to 1 cycle; Bit true: Yes; Gain: ~66.67%
 
**_Summary_**: Inline functions showed significant performance gains across all architectures,
 especially in cycle reductions, while noinline functions maintained consistent cycle counts with bit true results.
 
 
image

### CODE REVIEW

File: src/audio/tdfb/tdfb_direction.c [hunk 1]
```diff
@@ -285,9 +284,10 @@ static void level_update(struct tdfb_comp_data *cd, int frames, int ch_count, in
 	/* Calculate mean square level */
 	for (n = 0; n < frames; n++) {
 		s = *p;
+		tmp += ((int64_t)s * s);
```
Comment (lyakh) [inline src/audio/tdfb/tdfb_direction.c:287]: `s` is 16 bits, so casting it to 32 bits should be enough and should make this calculation faster. And yes, the external parentheses are redundant. And it actually doesn't matter where this calculation is made - here or where it was originally, right? So in its present form this change isn't improving anything

File: src/include/sof/audio/format_generic.h [hunk 2]
```diff
@@ -10,44 +10,37 @@
...
+}
+
+static inline int32_t sat_generic(int64_t x, int64_t max_val, int64_t min_val)
+{
+	int64_t mask_overflow = (max_val - x) >> 63;
+	int64_t mask_underflow = (x - min_val) >> 63;
+
+	/* Apply masks to selectively replace x with min or max values, or keep x as is. */
+	x = (x & ~mask_overflow) | (max_val & mask_overflow);
+	x = (x & ~mask_underflow) | (min_val & mask_underflow);
```
Comment (lyakh) [inline src/include/sof/audio/format_generic.h:42]: hm, I'm counting 12 64-bit operations here. Is this really faster than 2 `if`s?

**Review by ShriramShastry — COMMENTED** [review]
Thanks for your review. I have taken care for your suggestion `tmp += ((int32_t)s * s);` in [9169](https://github.com/thesofproject/sof/pull/9169).

File: src/include/sof/audio/format_generic.h (hunk 2 above)
Comment (lyakh) [reply src/include/sof/audio/format_generic.h:42]: @ShriramShastry sorry, I think there's a mistake in your analysis, at least in the first screenshot. In it you show the execution of the `sat_int32()` function which of course is much smaller now. But we need to look at the `sat_generic()` function which I'm pretty sure will be larger and possibly slower than the present version.

Comment (lgirdwood) [reply src/include/sof/audio/format_generic.h:42]: I think running a pipeline on real HW before and after will show the change in a E2E use case and show us an overall change

Comment (ShriramShastry) [reply src/include/sof/audio/format_generic.h:42]: Measurement done on `hifi5_ss_spfpu_7`

_PR GitHub code - Generic Saturation `Bitwise -32-bit` function_

image

_Original GitHub code - Generic Saturation `Branch -32` function_
image

_Xtensa setting_
image


_Profiler's Output_
image

_RSR.CCOUNT  -  Read Special Register_ 
image

Comment (lgirdwood) [reply src/include/sof/audio/format_generic.h:42]: @ShriramShastry sorry, I dont need the full simulator logs but, but just a abbreviated clip that shows before and after (as I cant tell what numbers I should be looking at above).

Comment (ShriramShastry) [reply src/include/sof/audio/format_generic.h:42]: OK,

In the image below, I captured performance data in terms of cycle count, and the profiler will return similar results.

Previously, the Original Function in GitHub used the branching `if-else` approach to calculate saturation. After that, the modified function under review calculates saturation using `bitwise` arithmetic.

_**(1) Generic_HiFi5**_

image


_**(2) Generic_HiFi4**_

image

_**(3) Generic_HiFi3**_

image

_**(4) ACE_3**_

image

_**(5) ACE_1**_

image

**Observation** : 

The saturation function limits the input value to a certain range, specific to the bit width of the output (32-bit, 24-bit, 16-bit, 8-bit). The `Cycles (Before)` and `Cycles (After)` columns indicate the number of cycles it took to execute the function before and after some modification, respectively.

Here's a summary of the cycles before and after the modification:

For `sat_int32`, the number of cycles remained the `same` before and after the modification, with 20 cycles for each operation.
For `sat_int24`, the number of cycles `increased` after the modification, from 16 cycles before to 21 cycles after.
For `sat_int16`, the number of cycles also `increased` after the modification, from 13 cycles before to 21 cycles after.
For `sat_int8`, the number of cycles `increased` after the modification as well, from 13 cycles before to 20 cycles after.

In summary, the modification resulted in an `increase` in the number of cycles for `sat_int24`, `sat_int16`, and `sat_int8 `functions, while the `sat_int32` function's performance remained `unchanged` in terms of cycles.

Comment (lyakh) [reply src/include/sof/audio/format_generic.h:42]: > In summary, the modification resulted in an `increase` in the number of cycles for `sat_int24`, `sat_int16`, and `sat_int8 `functions, while the `sat_int32` function's performance remained `unchanged` in terms of cycles.

Thank you, this matches my understanding, that those 64-bit calculations cannot result in better performance than a single comparison. Maybe this PR should be closed as a failed optimisation attempt

**Review by lyakh — CHANGES_REQUESTED** [review]
optimisation attempt seems to have failed

**Comment by ShriramShastry** [comment]
> optimisation attempt seems to have failed

With latest modification,  the gap in performance is addressed i.e. Before and After the changes, there's no performance difference

_**ACE-3**_

`There's gain of 1 cycle count. i.e. sat_int32: decreased from 21 to 20 cycles`

image

_**Generic HiFi5**_ 

image


_**Generic HiFi4**_ 

image


_**Generic HiFi3**_ 

image

**Comment by lgirdwood** [comment]
> There's gain of 1 cycle count. i.e. sat_int32: decreased from 21 to 20 cycles

@ShriramShastry  this is the sort of data we need for every change as its not really clear from all the table (there is too much data).  We just need to know the before and after in a simplified form like your comment above. 

> Here's a summary of the cycles before and after the modification :
```
sat_int32: No change (20 cycles both before and after)
sat_int24: Increased from 16 to 21 cycles
sat_int16: Increased from 13 to 21 cycles
sat_int8: Increased from 13 to 20 cycles
```
Ok, so this looks like a small gain for int32, but a bigger loss for 24, 16 and 8 and means the PR should not be merged until we can show no loss for any format. Do you have a plan to rework ?

**Comment by ShriramShastry** [comment]
> > There's gain of 1 cycle count. i.e. sat_int32: decreased from 21 to 20 cycles
> 
> @ShriramShastry this is the sort of data we need for every change as its not really clear from all the table (there is too much data). We just need to know the before and after in a simplified form like your comment above.
> 
> > Here's a summary of the cycles before and after the modification :
> 
> ```
> sat_int32: No change (20 cycles both before and after)
> sat_int24: Increased from 16 to 21 cycles
> sat_int16: Increased from 13 to 21 cycles
> sat_int8: Increased from 13 to 20 cycles
> ```
> 
> Ok, so this looks like a small gain for int32, but a bigger loss for 24, 16 and 8 and means the PR should not be merged until we can show no loss for any format. Do you have a plan to rework ?

I have addressed performance difference now !! With latest check-in there is no performance difference between `Before` and `After` .

**Comment by lyakh** [comment]
> I have reworked on sat-24,16 and 8 code and addressed performance loss !! With latest check-in there is no performance difference between `Before` and `After` my modification.

so, there's no need for this PR then. We don't make changes just because they don't make things worse. We only make changes if they make something better.

**Comment by ShriramShastry** [comment]
> > I have reworked on sat-24,16 and 8 code and addressed performance loss !! With latest check-in there is no performance difference between `Before` and `After` my modification.
> 
> so, there's no need for this PR then. We don't make changes just because they don't make things worse. We only make changes if they make something better.


Thank you for your feedback on the recent changes. I would like to address your concerns regarding performance.

**_Summary of Changes_**
The performance evaluation indicates that there is `no observable performance loss` due to the recent changes. Below is a summary of results from the performance evaluation:

image

**Benefits of the Optimization**
The aim of this optimization was not solely to enhance performance but also to enhance code reliability and maintainability. Here are the key benefits of the bitwise approach:

**Consistent Performance:**

There is no performance degradation; the cycle counts before and after the optimization remain the same across all tested values.

**Branch Prediction:**

The use of bitwise operations eliminates branches, which can reduce the risk of branch mispredictions and enhance predictability in pipeline performance.

**Code Clarity and Maintenance:**

The bitwise approach makes the code more compact and potentially less error-prone, improving readability and future maintainability.

**Future-proofing:**

Branch-free code is often favored in modern compiler and processor optimizations, which might yield further advantages on future architectures.

**Conclusion**
To summarize, the optimization has no negative impact on performance, maintains the same cycle counts as the original implementation, and provides additional long-term benefits in terms of code quality and predictability. Therefore, this change does not degrade performance and is a worthwhile improvement.

**Comment by marc-hb** [comment]
> To summarize, the optimization has no negative impact on performance, maintains the same cycle counts as the original implementation

Then you need to update the commit title and commit message. As they stand:

> This approach avoids branching, **improves efficiency**, and ensures
>  accurate saturation within specified bit-width limits.

Was there some accuracy issue that this is fixing?

This seems missing from the commit message too:
> which might yield further advantages on future architectures.

If this just a code clean-up  then say "more readable code" and don't mention performance at all. If it's more than a clean-up then you need some data or reference.

File: src/include/sof/audio/format_generic.h [hunk 3]
```diff
@@ -9,45 +11,85 @@
...
 static inline int32_t sat_int32(int64_t x)
 {
-	if (x > INT32_MAX)
-		return INT32_MAX;
-	else if (x < INT32_MIN)
-		return INT32_MIN;
-	else
-		return (int32_t)x;
+	int64_t mask_overflow = (INT32_MAX - x) >> 63;
+	int64_t mask_underflow = (x - INT32_MIN) >> 63;
```
Comment (marc-hb) [inline src/include/sof/audio/format_generic.h:27]: When `x` is big enough then `(x - INT32_MIN)` overflows, which is undefined behavior because it's signed.

Same overflow issue above when `x` is negative enough.

Signed overflow is undefined behavior:
https://wiki.sei.cmu.edu/confluence/display/c/INT32-C.+Ensure+that+operations+on+signed+integers+do+not+result+in+overflow

Code like this is being banned from the Linux kernel right now for security reasons:
https://lwn.net/Articles/979747/

There are compiler flags to define it but that would introduce a very subtle and hard to track dependency. For what benefit?
https://www.gnu.org/software/c-intro-and-ref/manual/html_node/Signed-Overflow.html


Even ignoring undefined behavior, this code is much, much harder to follow than the previous one.

Can branch prediction ever matter for code that small?

Comment (ShriramShastry) [reply src/include/sof/audio/format_generic.h:27]: > Can branch prediction ever matter for code that small?

Yes-  Please take a look at the low-level impacts through assembly comparision

The bitwise approach uses fewer branching instructions, making it more efficient as branch prediction may fail and cause pipeline stalls.

image

**_Bitwise Approach_**: Uses fewer branches, reducing the risk of pipeline stalls and improving efficiency.
**_Complexity_**: Despite appearing complex, bitwise operations provide stability and avoid undefined behavior from signed overflows.
**_Branch Prediction Impact_**: Even small code sections can significantly benefit from reduced branching

Comment (lgirdwood) [reply src/include/sof/audio/format_generic.h:27]: @ShriramShastry agree this should improve branch prediction and instruction pipeline throughput, but I think its best to show a real world test i.e. take a piece of FW code that is a heavy user of these APIs and place some timestamps around usage. The timestamps can then be measured before/after the changes with a real workload to show improvement.

**Review by lgirdwood — COMMENTED** [review]
Real world workload would show performance change and settle discussion.

File: src/include/sof/audio/format_generic.h (hunk 3 above)
Comment (ShriramShastry) [reply src/include/sof/audio/format_generic.h:27]: Sure

Comment (marc-hb) [reply src/include/sof/audio/format_generic.h:27]: > Complexity: Despite appearing complex, bitwise operations provide stability and avoid undefined behavior from signed overflows.

I don't think you understand what "undefined behavior" means. This is not a problem that shows up in assembly code.

Comment (ShriramShastry) [reply src/include/sof/audio/format_generic.h:27]: Thank you for your feedback. Marc.

I thank your attention to the complexity and stability provided by bitwise operations. However, I believe there might be a misunderstanding regarding the definition and implications of "undefined behavior" (UB) in the context of signed integer overflow.

**Explanation:**
Undefined Behavior (UB):

I went over the links shared above, In C and C++, "undefined behavior" refers to a situation where the language specification does not define what should happen. This can lead to unpredictable results as the compiler is free to generate any code for UB constructs.
A classic case of UB in C is signed integer overflow. For example, adding two large `int32_t` values that exceed the range of `int32_t` can result in UB because the C standard does not define the outcome of overflow for signed integers.

**Bitwise Operations:**
The bitwise operations used in the provided code ensure that values are clamped/bounded within their respective ranges without causing overflow, thereby ensuring predictable behavior.

For example, the code for sat_int32:

```
static inline int32_t sat_int32(int64_t x) {
    int64_t mask_overflow = (INT32_MAX - x) >> 63;
    int64_t mask_underflow = (x - INT32_MIN) >> 63;

    x = (x & ~mask_overflow) | (INT32_MAX & mask_overflow);
    x = (x & ~mask_underflow) | (INT32_MIN & mask_underflow);
    return (int32_t)x;
}
```

Here, bitwise operations are used to safely bounded `int64_t` values to the `int32_t `range, avoiding the issues of signed overflow.

**Understanding in Assembly:**
While undefined behavior often has unpredictable and compiler-dependent outcomes, bitwise operations tend to translate directly to assembly without causing UB.
Ensuring signed overflows are handled in C code avoids reliance on compiler-specific behavior and promotes consistent, predictable execution across different platforms.

**Summary:**
Bitwise operations provide a robust way to handle out-of-range values, preventing undefined behavior stemming from signed integer overflow in C. This ensures predictable, stable, and portable code execution.

If there is an aspect I might have overlooked regarding the specific implications on assembly code or if you meant something different by undefined behavior, please let me know, and I can address it further.

**Comment by singalsu** [comment]
I did a performance comparison with git main vs main + #9170 with TGL build with gcc and xcc. The gcc build currently can't run enabled DRC (reason unknown, a regression or just too much load) so I used different configurations for hw:0,0 playback on sof-hda-generic-4ch.tplg

GCC (gain 1.1 44, gain 2.1 44, EQIIR eq_iir_flat.txt, EQFIR eq_fir_flat.txt, DRC passthrough.txt  off)
- main 69.8 MCPS
- 9170 74.2 MCPS

XCC (gain 1.1 44, gain 2.1 44, EQIIR eq_iif_spk.txt, EQFIR eq_fir_spk.txt, DRC  threshold_-25_knee_15_ratio_10.txt on)
- main 96.8 MCPS
- 9170 96.8 MCPS

I'll see if I can repeat this with some remote MTL device. But this doesn't look for TGL.

**Comment by singalsu** [comment]
Similar result with MTL, no improvement with xcc build,  and about same 4 MCPS slower with gcc build of this PR.

File: src/include/sof/audio/format_generic.h (hunk 3 above)
Comment (marc-hb) [reply src/include/sof/audio/format_generic.h:27]: > The bitwise operations used in the provided code ensure that values are clamped/bounded within their respective ranges without causing overflow, thereby ensuring predictable behavior.
```C
static inline int32_t sat_int32(int64_t x) {
    int64_t mask_overflow = (INT32_MAX - x) >> 63;
```

This is not true because the shift happens _after_ the undefined behavior.

Comment (marc-hb) [reply src/include/sof/audio/format_generic.h:27]: > > Can branch prediction ever matter for code that small?

> Yes- Please take a look at the low-level impacts through assembly comparison

Assembly differences are interesting but they're not answering the question. The question was "Does this matter? (and how much)". For this sort of optimization to matter you need:

1. A branch predictor
2. A branch prediction which is difficult to make and often enough wrong
3. A long enough pipeline
4. Code in the critical path

Most performance optimizations tend to be very "cruel" because they're correct in theory and they don't matter in practice. That's what Donald Knuth meant when he wrote "premature optimization is the root of all evil" https://en.wikiquote.org/wiki/Donald_Knuth

**Comment by andyross** [comment]
So... I hate to say it but I'm pretty sure this change is just a big noop to the optimizer.  Here's some test code that extracts the two variants in individually compilable form, and a just-non-trivial-enough outer loop to let the compiler try to use the code in a natural setting.  And the generated code from xt-clang (RI-2022.10, at -Os) is essentially identical.  Both variants reduce to SALTU instructions.  There is no branching in the current code nor masking in the new one; the compiler is smart enough to recognize what's happening.

```c
#define INT_MAX 0x7fffffff
#define INT_MIN 0x80000000

static inline int sat32_cmp(long long x)
{
	return x > INT_MAX ? INT_MAX : (x < INT_MIN ? INT_MIN : x);
}

static inline int sat32_mask(long long x)
{
	long long mask_overflow = (INT_MAX - x) >> 63;
	long long mask_underflow = (x - INT_MIN) >> 63;

	x = (x & ~mask_overflow) | (INT_MAX & mask_overflow);
	x = (x & ~mask_underflow) | (INT_MIN & mask_underflow);
	return (int) x;
}

void sat_array_add_cmp(int *dst, int *a, int *b, int n)
{
	for (int i = 0; i < n; i++) {
		dst[i] = sat32_cmp(a[i] + (long long) b[i]);
	}
}

void sat_array_add_mask(int *dst, int *a, int *b, int n)
{
	for (int i = 0; i < n; i++) {
		dst[i] = sat32_mask(a[i] + (long long) b[i]);
	}
}
```

**Comment by andyross** [comment]
And indeed, gcc[1] is **not** smart enough to figure it out, and emits the code more or less as written.  But... honestly my read is that the branch-based code is better and not worse.  It's a lot fewer instructions, and Xtensa pipelines are really short in practice.  Certainly I've never bumped up against branch stalls as a performance case myself; most branches are very cheap.

[1] Also a clang I have built from the Espressif LLVM tree that I've been playing with.  It actually appears to build Zephyr for SOF targets pretty well, if anything slightly better code than gcc, and all the variants can go in one toolchain instead of N.  I need to get that cleaned up and submitted somewhere, with a writeup for Zephyr...

**Comment by singalsu** [comment]
I used for test these compiler versions

- TGL gcc: xtensa-intel_tgl_adsp_zephyr-elf-gcc (Zephyr SDK 0.16.4) 12.2.0
- MTL gcc: xtensa-intel_ace15_mtpm_zephyr-elf-gcc (Zephyr SDK 0.16.4) 12.2.0
- TGL xcc: xt-xcc version 12.0.8, RG-2017.8-linux, cavs2x_LX6HiFi3_2017_8
- MTL xcc:  XtensaTools-14.10 clang version 10.0.1, RI-2022.10-linux, ace10_LX7HiFi4_2022_10

Detailed resuls:

Screenshot from 2024-07-19 10-41-25

**Comment by singalsu** [comment]
Used algorithms settings for xcc

```
amixer cset name='Post Mixer Analog Playback Volume' 44
amixer cset name='Pre Mixer Analog Playback Volume' 44
sof-ctl -n 34 -s ~/ctl4/eq_iir_spk.txt 
sof-ctl -n 35 -s ~/ctl4/eq_fir_spk.txt 
sof-ctl -n 36 -s ~/ctl4/drc/threshold_-25_knee_15_ratio_10.txt
amixer cset name='Post Mixer Analog Playback DRC switch' on```

Settings for gcc

```
amixer cset name='Post Mixer Analog Playback Volume' 44
amixer cset name='Pre Mixer Analog Playback Volume' 44
sof-ctl -n 34 -s ~/ctl4/eq_iir_flat.txt 
sof-ctl -n 35 -s ~/ctl4/eq_fir_flat.txt 
sof-ctl -n 36 -s ~/ctl4/drc/passthrough.txt
amixer cset name='Post Mixer Analog Playback DRC switch' off```

(sorry, should use those long names instead of numids with sof-ctl, in TGL  and MTL the numids are not the same, the long string names would be...)

**Comment by marc-hb** [comment]
> TGL gcc: xtensa-intel_tgl_adsp_zephyr-elf-gcc (Zephyr SDK 0.16.4) 12.2.0
> MTL gcc: xtensa-intel_ace15_mtpm_zephyr-elf-gcc (Zephyr SDK 0.16.4) 12.2.0
> TGL xcc: xt-xcc version 12.0.8, RG-2017.8-linux, cavs2x_LX6HiFi3_2017_8
> MTL xcc: XtensaTools-14.10 clang version 10.0.1, RI-2022.10-linux, ace10_LX7HiFi4_2022_10
> 

You can get the front-end version with `-dumpversion`

```
RG-2017.8-linux/XtensaTools/bin/xt-xcc  -dumpversion (gcc frontend)
4.2.0

RI-2022.10-linux/XtensaTools/bin/xt-clang -dumpversion (clang frontend)
10.0.1
```

**Comment by lgirdwood** [comment]
@ShriramShastry I assume this PR has been split up and been re-created into smaller PRs ? if so, we should close and fous on the new PRs.

**Comment by lgirdwood** [comment]
> @ShriramShastry I assume this PR has been split up and been re-created into smaller PRs ? if so, we should close and focus on the new PRs.
